packages/dds/tree/src/simple-tree/discrepancies.ts

```diff
@@ -321,69 +491,55 @@ function* trackObjectNodeDiscrepancies(
-		switch (result.type) {
-			case "aExtra": {
-				if (result.value.kind === storedEmptyFieldSchema.kind) {
-					// In one of view/stored, this field is explicitly forbidden, but in the other it is implicitly forbidden
-					// (by way of omission). We treat these identically anyway.
-					break;
-				}
-				yield {
-					identifier,
-					fieldKey,
-					mismatch: "fieldKind",
-					view: result.value.kind,
-					stored: storedEmptyFieldSchema.kind,
-				} satisfies FieldKindDiscrepancy;
-				break;
-			}
-			case "bExtra": {
-				if (result.value.kind === storedEmptyFieldSchema.kind) {
-					// In one of view/stored, this field is explicitly forbidden, but in the other it is implicitly forbidden
-					// (by way of omission). We treat these identically anyway.
-					break;
-				}
-				yield {
-					identifier,
-					fieldKey,
-					mismatch: "fieldKind",
-					view: storedEmptyFieldSchema.kind,
-					stored: result.value.kind,
-				} satisfies FieldKindDiscrepancy;
-				break;
-			}
-			case "both": {
-				yield* getFieldDiscrepancies(result.valueA, result.valueB, identifier, fieldKey);
-				break;
-			}
-			default:
-				break;
+	const storedToSimpleKeys = new Map<FieldKey, string | symbol>();
+
+	for (const [key, { storedKey, schema }] of view) {
+		const storedSchema = stored.objectNodeFields.get(storedKey);
+		storedToSimpleKeys.set(storedKey, key);
+
+		// If the view schema has a field that's not in the stored schema
+		if (storedSchema === undefined) {
+			const viewKind =
+				convertFieldKind.get(schema.kind) ??
+				fail("A conversion from a FieldKind to a FlexFieldKind should exist");
+			yield {
+				identifier,
+				fieldKey: storedKey,
+				mismatch: "fieldKind",
+				view: viewKind.identifier,
+				stored: storedEmptyFieldSchema.kind,
+			} satisfies FieldKindDiscrepancy;
+		} else {
+			yield* getFieldDiscrepancies(
+				schema,
+				storedSchema,
+				identifier,
+				storedKey,
+				viewKindIsSequence,
+			);
 		}
 	}
-}
 
-function* compareMaps<K, V1, V2>(
-	a: ReadonlyMap<K, V1>,
-	b: ReadonlyMap<K, V2>,
-): Iterable<
-	| { type: "aExtra"; key: K; value: V1 }
-	| { type: "bExtra"; key: K; value: V2 }
-	| { type: "both"; key: K; valueA: V1; valueB: V2 }
-> {
-	for (const [key, valueA] of a) {
-		const valueB = b.get(key);
-		yield valueB === undefined
-			? { type: "aExtra", key, value: valueA }
-			: { type: "both", key, valueA, valueB };
-	}
-	for (const [key, valueB] of b) {
-		if (!a.has(key)) {
-			yield { type: "bExtra", key, value: valueB };
+	for (const [fieldKey, schema] of stored.objectNodeFields) {
+		const viewSchema = storedToSimpleKeys.get(fieldKey);
+		// If the stored schema has a field that's not in the view schema
+		if (viewSchema === undefined) {
+			if (schema.kind === storedEmptyFieldSchema.kind) {
+				// In one of view/stored, this field is explicitly forbidden, but in the other it is implicitly forbidden
+				// (by way of omission). We treat these identically anyway.
+				continue;
```

Comment:
I don't think this is a good way to do this check, nor a good explanation.

If you want to handle forbiddenFieldKindIdentifier, refer to it directly, not via storedEmptyFieldSchema.kind, since in the future we might change storedEmptyFieldSchema.kind. I tcould be optional for example, and that would still be empty, but would break this code.

Additionally, a forbidden field must be empty, but a field not mention in an object view schema might be essentially be any optional field it allow unknown optional fields is set so I would not call that implicitly forbidden.

```suggestion
			if (schema.kind === forbiddenFieldKindIdentifier.kind) {
				// In the stored schema the field is explicitly forbidden.
				// This has the same semantics of the field not being mentioned in the stored schema,
				// and thus is compatible with the view schema which does not mention this field.
				continue;
```